lib/public_key/src/pubkey_ocsp.erl

```diff
@@ -153,27 +156,80 @@ get_nonce_value([#'Extension'{
 get_nonce_value([_Extn | Rest]) ->
     get_nonce_value(Rest).
 
-verify_ocsp_signature(ResponseDataDer, SignatureAlgo, Signature,
-                      Certs, ResponderID) ->
-    case find_responder_cert(ResponderID, Certs) of
+verify_signature(ResponseDataDer, SignatureAlgo, Signature,
+                      Certs, ResponderID, IssuerCert) ->
+    case find_responder_cert(ResponderID, Certs, IssuerCert) of
         {ok, Cert} ->
-            do_verify_ocsp_signature(
+            do_verify_signature(
                 ResponseDataDer, Signature, SignatureAlgo, Cert);
         {error, Reason} ->
             {error, Reason}
     end.
 
-find_responder_cert(_ResponderID, []) ->
+verify_past_timestamp(Timestamp) ->
+    {Now, TimestampSec} = get_time_in_sec(Timestamp),
+    verify_timestamp(Now, TimestampSec, past_timestamp).
+
+verify_future_timestamp(Timestamp) ->
+    {Now, TimestampSec} = get_time_in_sec(Timestamp),
+    verify_timestamp(Now, TimestampSec, future_timestamp).
+
+verify_timestamp(Now, Timestamp, past_timestamp) when Timestamp =< Now ->
+    ok;
+verify_timestamp(Now, Timestamp, future_timestamp) when Now =< Timestamp ->
+    ok;
+verify_timestamp(_, _, _) ->
+    {error, ocsp_stale_response}.
+
+get_time_in_sec(Timestamp) ->
+    Now = calendar:datetime_to_gregorian_seconds(calendar:universal_time()),
+    TimestampSec = pubkey_cert:time_str_2_gregorian_sec(
+                     {generalTime, Timestamp}),
+    {Now, TimestampSec}.
+
+verify_next_update(asn1_NOVALUE) ->
+    ok;
+verify_next_update(NextUpdate) ->
+    verify_future_timestamp(NextUpdate).
+
+find_responder_cert(_ResponderID, [], _) ->
     {error, ocsp_responder_cert_not_found};
-find_responder_cert(ResponderID, [Cert | TCerts]) ->
-    case is_responder(ResponderID, Cert) of
+find_responder_cert(ResponderID, [Cert | TCerts], IssuerCert) ->
+    maybe
+        true ?= is_responder_cert(ResponderID, Cert),
+        ok ?= verify_responder_cert(Cert, IssuerCert),
+        {ok, Cert}
+    else
+        _-> find_responder_cert(ResponderID, TCerts, IssuerCert)
+    end.
+
+verify_responder_cert(Cert, IssuerCert) ->
+    Case1 =
+        %% the CA who issued the certificate in question signed the
+        %% response
+        fun() -> Cert == IssuerCert end,
+    Case2 =
+        %% a CA Designated Responder (Authorized Responder, defined in
+        %%      Section 4.2.2.2) who holds a specially marked certificate
+        %%      issued directly by the CA, indicating that the responder may
+        %%      issue OCSP responses for that CA (id-kp-OCSPSigning)
+        fun() -> public_key:pkix_is_issuer(Cert, IssuerCert) andalso
+                                 designated_for_ocsp_signing(Cert) end,
+    Case3 =
+        %% a Trusted Responder whose public key is trusted by the requestor
+        fun() ->
+                Pred = fun(#cert{otp = CACert}) -> IssuerCert == CACert end,
+                lists:any(Pred, public_key:cacerts_get())
```

Comment:
We can not use cacerts_get/0 here, we can not assume who the user wants to trust. cacerts_get/0 is typically who webbrowser would trust.  It needs to be configurable.  The ssl application would default to its cacerts option. In the long run we might want to have a special stapling trusted option, the one we agreed that we could skip for now to reduce complexity.  Also I think verify responder cert is a confusing name as you do not verify the signature in this function, this is done later this is still about finding the potential issuer.  This code path may also be problematic if there exist more than one potential issuer in the set of trusted certs!

Reply:
I've applied this comment apart from passing trusted certs from ssl to public_key, which I would like to discuss first.

Reply:
this is not realized in `ssl_certificate:verify_cert_extensions/4` which provides public key with IsTrustedResponderFun

resolving